File: Data Structures & Algorithms/valid-binary-search-tree/submission-15.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        min_left, min_right = float('-inf'), float('inf')

        def dfs(node, l, r):
            if not node:
                return True
            if node.val <= l or node.val >= r:
                return False
            
            return dfs(node.left, l, node.val) and dfs(node.right, node.val, r)
        
        return dfs(root, min_left, min_right)

File: Data Structures & Algorithms/depth-of-binary-tree/submission-15.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    pub fn max_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let root_node = match root {
            Some(node) => node,
            None => return 0,
        };

        let (left, right) = {
            let node = root_node.borrow();
            (node.left.clone(), node.right.clone())
        };

        let left_depth = Self::max_depth(left);
        let right_depth = Self::max_depth(right);

        1 + left_depth.max(right_depth)
    }
}
